CMakeLists.txt

```diff
@@ -1,6 +1,10 @@
 cmake_minimum_required(VERSION 3.6)
 project(mlpack C CXX)
 
+#Replace all CMAKE_SOURCE_DIR and CMAKE_BINARY_DIR with MLPACK_ prefix.
+set(MLPACK_SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR} CACHE STRING "Source path of mlpack")
+set(MLPACK_BINARY_DIR ${CMAKE_CURRENT_BINARY_DIR} CACHE STRING "Binary path of mlpack")
```

Comment:
Does it need to be a cache variable?  I had thought that cache variables were really meant to be user-configurable, but if the user specifies something random for these, I don't think it will work very well---`MLPACK_SOURCE_DIR` and `MLPACK_BINARY_DIR` need to be the root of the mlpack CMake configuration.